design/baremetal-operator/raid-api.md

```diff
@@ -0,0 +1,200 @@
+   when provisioning a host.
+2. Outline the controller/provisioner logic changes needed to pass the
+   instructions to Ironic to have the storage configured.
+3. Support basic RAID levels `0`, `1`, and `1+0`.
+
+### Non-Goals
+
+1. Provide default RAID configuration.
+2. Retain existing RAID configuration on a host if it matches the
+   desired configuration at the time of provisioning (we make no
+   assumptions about preserving the state of the storage system or its
+   contents).
+3. Re-provision a host if the RAID settings change. (Users are
+   expected to explicitly deprovision the host and then reprovision
+   it.)
+4. Support every possible RAID configuration.
+
+## Proposal
+
+### User Stories
+
+1. As a user, I can specify the RAID configuration for a host before
+   provisioning it.
+2. As a user, I can choose between hardware and software RAID for each
+   host.
+
+## Design Details
+
+### API Changes
+
+The `BareMetalHost` API will change to accept RAID configuration
+inputs and to reflect the settings that were used when a host is
+provisioned (in case they are changed after provisioning).
+
+New `spec` fields will be added in a section called `raid`, with
+separate fields describing the hardware and software RAID settings. If
+the hardware settings are provided, the software settings are ignored.
+
+For hardware RAID, the inputs describe how to combine physical disks
+into a RAID volume via the RAID controller hardware in the host.
+
+```yaml
+spec:
+  raid:
+    hardwareVolumes:
+      - sizeGibibytes: 1000
+        level: 1
+        name: volume-name
+        rotational: true
+        numberOfPhysicalDisks: 2
+```
+
+- The RAID level for a volume is required and should be expressed as
+  one of `"1"`, `"0"`, or `"1+0"`.
+- The size for a volume is optional and should be expressed in
+  gibibytes.
+- The number of disks to use for the volume defaults to a minimum
+  value based on the RAID level.
+
+For software RAID, the inputs describe how to configure the operating
+system to see physical disks in a RAID configuration through software
+drivers, without using any special hardware.
+
+```yaml
+spec:
+  raid:
+    softwareVolumes:
+      - sizeGibibytes: 1000
+        level: 1
+        physicalDisks:
+          - deviceName: /dev/sda
+          - deviceName: /dev/sdb
+```
+
+- The RAID level for a volume is required and should be expressed as
+  one of `"1"`, `"0"`, or `"1+0"`.
+- The size for a volume is optional and should be expressed in
+  gibibytes.
+- The disks to use are specified using the same syntax as the
+  `rootDeviceHints`.
+
+Software RAID is constrained in other ways:
+
+- The first volume is always used as the root volume.
+- There must be 0, 1, or 2 software RAID devices.
+- The first software RAID volume must use RAID level 1 to ensure the
+  host boots in the event of a disk failure.
+- The second software RAID volume may use RAID level `0`, `1`, or
+  `1+0`.
```

Comment:
I don't see anywhere in the implementation that enforces the rest of these rules, either.

Reply:
```	// +kubebuilder:validation:Enum="0";"1";"1+0"
	Level string `json:"level" required:"true"`
```
Can it not ensure the rule?

Reply:
Yes, the API validation will enforce the enum values.

Reply:
Yes, and ironic enforces this rule in it's implementation of software RAID, so we're good to go.

Reply:
If we wait for Ironic to enforce all of the rules, then the user may wait 30 minutes to receive an error message. I think we're going to want to anticipate those sorts of failures and report problems early, before we even attempt to provision.

Reply:
And I presume this kube builder verification will be able to point those inconsistencies out? Or what are we missing here?

Reply:
The kubebuilder validations should be good enough. I just wanted to point out that we need to do our own validation as much as possible.

Reply:
I see. I'm sure we're doing some validation in the implementation as well.